addons/zylann.hterrain/hterrain_detailer.gd

```diff
@@ -0,0 +1,154 @@
+tool
+
+class DetailLayerSorter:
+	static func sort_descending(a, b):
+		if a.view_distance > b.view_distance:
+			return true
+		return false
+	
+const Logger = preload("./util/logger.gd")
+
+const CHUNK_SIZE = 32
+const AABBS_CACHE_SIZE = 256
+
+var _terrain = null
+var _detail_layers := []
+
+# Vector2 => AABB
+var _cached_aabbs := {}
+
+var _logger := Logger.get_for(self)
+
+
+func set_terrain(terrain):
+	assert(terrain is Spatial)
+	_terrain = terrain
+
+func update_materials():
+	for layer in _detail_layers:
+		layer.update_material()
+
+func add_map(index: int):
+	for layer in _detail_layers:
+		# Shift indexes up since one was inserted
+		if layer.layer_index >= index:
+			layer.layer_index += 1
+		layer.update_material()
+
+func remove_map(index: int):
+	for layer in _detail_layers:
+		# Shift indexes down since one was removed
+		if layer.layer_index > index:
+			layer.layer_index -= 1
+		layer.update_material()
+
+func set_transform(terrain_transform: Transform):
+	update_materials()
+	# Clear cached aabbs because the transform changed
+	_cached_aabbs.clear()
+		
+func process(delta: float, viewer_pos: Vector3):
+	if _terrain == null:
+		return
+
+	var local_viewer_pos = _terrain.global_transform.affine_inverse() * viewer_pos
+	
+	var viewer_cx = local_viewer_pos.x / CHUNK_SIZE
+	var viewer_cz = local_viewer_pos.z / CHUNK_SIZE
+
+	var map_res = _terrain.get_data().get_resolution()
+	var map_scale = _terrain.map_scale
+
+	var terrain_size_x = map_res * map_scale.x
+	var terrain_size_z = map_res * map_scale.z
+
+	var terrain_chunks_x = terrain_size_x / CHUNK_SIZE
+	var terrain_chunks_z = terrain_size_z / CHUNK_SIZE
+	
+
+	var visible_layers = get_layers()
+	for idx in range(visible_layers.size() -1, -1, -1):
+		var layer = visible_layers[idx]			
+		if !layer.visible:
+			visible_layers.remove(idx)
+	if !visible_layers.empty():		
+		visible_layers.sort_custom(DetailLayerSorter, "sort_descending")
+	
+	var aabbs_updated = false
+	for layer in visible_layers:
+		var cr = int(layer.view_distance) / CHUNK_SIZE + 1
+	
+		var cmin_x = viewer_cx - cr
+		var cmin_z = viewer_cz - cr
+		var cmax_x = viewer_cx + cr
+		var cmax_z = viewer_cz + cr
+	
+		if cmin_x < 0:
+			cmin_x = 0
+		if cmin_z < 0:
+			cmin_z = 0
+		if cmax_x > terrain_chunks_x:
+			cmax_x = terrain_chunks_x
+		if cmax_z > terrain_chunks_z:
+			cmax_z = terrain_chunks_z
+			
+		var x_range = range(cmin_x, cmax_x)
+		var z_range = range(cmin_z, cmax_z)
+
+		# Only updates AABBs once for the detail layer with the biggest view distance, and the result is then shared between all layers 		
+		if !aabbs_updated:
```

Comment:
I think now that there is a cache, there is no point updating AABBs every frame even. Most of the time this is going to recalculate the same values, with the cost of iterations and hashing into the dictionary added on top. Only newly introduced chunks will have values to calculate, and this doesn't even happen every frame. The original code was doing the calculations on the fly, which is why this part was done regularly.

I believe it would be possible to compute AABBs for ALL chunks of the terrain just once and have them in a grid (not a dict), and only update them when the terrain changes. That's already done inside `HTerrainData`, where `_chunked_vertical_bounds` IS a similar cache, but for the entire terrain and with a smaller chunk size. This is why HTerrainDetailLayers were computing their own based on it. This was supposed to be fast if done in C++, but GDScript be GDScript. 
There might even be a way to do optimize this without a cache, but GDScript heavily penalizes the amount of code and calls (which in C++ can be optimized out at compile time)

Reply:
I updated the code a bit. Now I compute AABBs for all detail chunks once and stores the results in the dict, and only updates them when (1)terrain transform changes or (2)terrain chunks rebuild